### DISCUSSION TITLE
WS doesn't connect

### DISCUSSION THREAD

**Discussion by phbou72** [ElMassimo/vite_ruby#327] [Q&A] [answered]
Hi! I'm getting those errors when trying to make Vite + Ruby work : 
`client.ts:78 WebSocket connection to 'wss://an-host.io:3036/vite-dev/' failed: `
`an-host:3036/vite-dev/ net::ERR_CONNECTION_CLOSED`

The assets are loading but no HMR. My local dev server is hosted on https. Not sure if it impact the connection for the WS. However, I also can't connect on https://127.0.0.1:3036/vite-dev/. I'm met with a 404 error.

Any idea what could be the problem?

**Here's my configurations** 
_vite.config.ts_
```
import {defineConfig} from 'vite';
import RailsPlugin from 'vite-plugin-rails';
import react from '@vitejs/plugin-react-swc';
import tsconfigPaths from 'vite-tsconfig-paths';

export default defineConfig({
  plugins: [
    RailsPlugin(),
    react(),
    tsconfigPaths(),
  ],
  base: '/',
  css: {
    modules: {
      scopeBehaviour: 'local',
      globalModulePaths: [new RegExp('app/ui')],
      localsConvention: 'camelCaseOnly',
    },
  },
});
```

_vite.json_
```
{
  "all": {
    "sourceCodeDir": "app/ui",
    "watchAdditionalPaths": []
  },
  "development": {
    "autoBuild": true,
    "publicOutputDir": "vite-dev",
    "port": 3036
  },
  "test": {
    "autoBuild": true,
    "publicOutputDir": "vite-test",
    "port": 3037
  }
}
```

_application.html.erb_
```
<!DOCTYPE html>
<html>
  <head>
    <title>Vite+Ruby</title>
    <meta charset="UTF-8" />
    <%= csrf_meta_tags %>
    <%= csp_meta_tag %>

    <%= vite_client_tag %>
    <%= vite_react_refresh_tag %>
    <%= vite_typescript_tag 'application.tsx' %>
  </head>

  <body>
    <div id="app" />

    <%= yield %>
  </body>
</html>
```

_routes.rb__
```
Rails.application.routes.draw do
  ...
  root "welcome#index"
  get '*path', to: "welcome#index"
end
...

**Comment by phbou72** [comment]
Here's a demo repo you can pull  : https://github.com/phbou72/vite-ruby-test

**Comment by phbou72** [comment]
Is this issue related to this ? https://vite-ruby.netlify.app/guide/troubleshooting.html#making-hmr-work-with-puma-dev-in-https-apps

**Comment by ElMassimo** [comment, marked as answer]
> My local dev server is hosted on https

In that case, you'll need to configure Vite dev server to use [`https`](https://vitejs.dev/config/server-options.html#server-https) by providing a valid certificate.

You will also need to set [`"https": true`](https://vite-ruby.netlify.app/config/index.html#https) so that Vite Ruby is aware, or use the [`skipProxy`](https://vite-ruby.netlify.app/config/index.html#skipproxy-experimental) experimental setting so that assets are requested from the Vite dev server directly.

**Comment by phbou72** [comment]
If I tried this config with the plugin @vitejs/plugin-basic-ssl: 
```
import {defineConfig} from 'vite';
import RailsPlugin from 'vite-plugin-rails';
import react from '@vitejs/plugin-react-swc';
import tsconfigPaths from 'vite-tsconfig-paths';
import basicSsl from '@vitejs/plugin-basic-ssl';

export default defineConfig({
  plugins: [basicSsl(), RailsPlugin(), react(), tsconfigPaths()],
  base: '/',
  css: {
    modules: {
      scopeBehaviour: 'local',
      globalModulePaths: [new RegExp('app/ui')],
      localsConvention: 'camelCaseOnly',
    },
  },
  server: {
    cors: {
      origin: '*',
    },
    https: true,
    hmr: {
      protocol: 'wss',
    },
  },
});
```

I end up with this kind of error: https://app.io/vite-dev/@vite/client net::ERR_ABORTED 500

However, if I set `server.https = false` and `server.hrm.protocol = 'ws'` the app load but not the HMR : 
`[vite] failed to connect to websocket (SecurityError: Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.).`


Feel like I'm missing something?

**Reply by ElMassimo** [reply]
> Failed to construct 'WebSocket': An insecure WebSocket connection may not be initiated from a page loaded over HTTPS.

That makes sense, updated my answer to remove this as it's not a viable option, and added a mention of `skipProxy` which might be useful in this scenario.

Make sure that you configure the `host` correctly as well. What does `app.io` resolve to in your configuration, `localhost`?

**Reply by phbou72** [reply]
OK, so I figured out the right config : 

<img width="979" alt="Screenshot 2023-01-16 at 12 34 02 PM" src="https://user-images.githubusercontent.com/1980396/212737351-11827544-e0cb-43df-9743-a593065b11ce.png">


This work for me! Thanks for your help @ElMassimo !
